Rails 5 update

sonalkr132 opened the pull request:
Please read commit message for explanation of changes.

It is very likely that I may have missed deprecations which didn't show up in test, feel free to open issues when you see them.

I have removed [`quite_assets`](https://github.com/evrone/quiet_assets) gem because their rails requirement wouldn't have allowed update and looks like gem is abandoned. The feature is now [implemented in sprocket-rails](https://github.com/rails/sprockets-rails/pull/355), we just have to be wait for version release.

I really want to compare before and after performance, object allocation etc. I would really appreciate it if someone pointed me to relevant resources or took out time for general advice.

dwradcliffe commented:
I haven't looked at the commits yet, but if there are any of these things that are compatible with both Rails 5 and the current version we should get those into separate PRs and merge them sooner. Smaller changes are always better to test.

cc: @arthurnn @rafaelfranca

sonalkr132 commented:
> if there are any of these things that are compatible with both Rails 5 and the current version we should get those into separate PRs and merge them sooner.

I am not sure if that would be a good idea. You won't be able to see anything in the logs other than deprecation warnings.

This PR is not as bad it looks. Just gemfiles change amount for 128 lines, `id: ..` => `params: { id: .. }` cloc at 138 lines and adding version to migration is 88 line (82% total change).

homu commented:
:umbrella: The latest upstream changes (presumably #1327) made this pull request unmergeable. Please resolve the merge conflicts.

sonalkr132 commented:
Thank you for checking out the PR @kmcphillips :octocat:

>  you need to add config.assets.quiet = true to environments/development.rb for the equivalent behaviour.

I didn't add it because I didn't want to add a configuration which did nothing at the point(commit) it was added. `quite_assets` isn't anything critical as in we can live without it for a day or two. I hope it makes sense :sweat_smile:

kmcphillips commented:
#1352 simplifies this a bit. `sprockets-rails` 3.1.0 was released today.

sonalkr132 commented:
I have great new.. :disappointed: 

It has treading issue and as they go.. it is not reliably reproduceable. Failing [build1](https://travis-ci.org/sonalkr132/rubygems.org/jobs/140843290#L922), [build2](https://travis-ci.org/rubygems/rubygems.org/jobs/140831890#L1126), [build3](https://travis-ci.org/sonalkr132/rubygems.org/jobs/140841700#L925).
Almost all of them are on [`increment` function](https://github.com/rubygems/rubygems.org/blob/master/app/models/gem_download.rb#L45-L52) of GemDownload. How do I even report this?

homu commented:
:umbrella: The latest upstream changes (presumably f32321d) made this pull request unmergeable. Please resolve the merge conflicts.

sonalkr132 commented:
I have updated this to 5.0.0. Treading error was fixed here: https://github.com/rails/rails/pull/25707
I guess we will have to wait for 5.0.1.

prathamesh-sonpatki commented on config/application.rb:
```diff
@@ -28,6 +25,9 @@ class Application < Rails::Application
+
+    # Do not halt callback chains when a callback returns false.
+    ActiveSupport.halt_callback_chains_on_return_false = false
   end
```
@sonalkr132 Did you run `rails app:update`, this config will go to `config/initializers/new_framework_defaults.rb` after that.

sonalkr132 replied:
I was hoping I would change config after we are sure that changes up to this point don't break anything.
However, now it seems like it is going to be a while before this gets deployed.. I have added the changes of `rails app:update`. Thanks for having a look at the PR :)

homu commented:
:umbrella: The latest upstream changes (presumably e0006da) made this pull request unmergeable. Please resolve the merge conflicts.

simi commented on Gemfile:
```diff
@@ -53,9 +53,7 @@ group :development, :test do
-group :development do
-  gem 'rails-erd'
-end
+gem 'rails-erd', group: :development
```
Is this needed?

sonalkr132 replied:
It is used by https://github.com/rubygems/rubygems.org/blob/071df7d625ee18be64d225e78abb445fdbf17462/lib/tasks/gen_erd.rake

simi replied:
Sure, but why changed?

sonalkr132 replied:
Oh! It has to do with rebase :stuck_out_tongue_closed_eyes: I had removed `quite_assets` changed group to one line but `quite_assets` was removed in different PR and I had to rebase.

I guess it saves two lines, but yes, the change is unnecessary. May be I can restore it when I update the PR for 5.0.1 if you think it is important.

simi replied:
Since this PR is changing 169 files I think it will be nice to keep it as minimal as it is possible for future reviewers.

simi commented:
I'm not sure if `ApplicationRecord` is needed and it will save a lot of line changes since it is not used (it is just blank abstract class). Is it really worth it now?

simi commented on app/controllers/profiles_controller.rb:
```diff
@@ -5,10 +5,8 @@ def edit
-    @user           = User.find_by_slug!(params[:id])
-    rubygems        = @user.rubygems_downloaded
-    @rubygems       = rubygems.slice!(0, 10)
-    @extra_rubygems = rubygems
```
Is this removal of `@extragems_rubygems` and `rubygems` var assignments and following view change related to Rails 5 update or is it just a code cleanup/refactoring?

simi commented on config/initializers/new_framework_defaults.rb:
```diff
@@ -0,0 +1,23 @@
+Rails.application.config.active_record.belongs_to_required_by_default = false
+
+# Do not halt callback chains when a callback returns false. Previous versions had true.
+ActiveSupport.halt_callback_chains_on_return_false = false
```
This is the only config option not respecting old versions and also resulting in code changes. Do we really need it right now?

simi commented:
I left some notes @sonalkr132 to your awesome work!.
Anyway I'm just random passerby so it will be good to wait for someone from RubyGems.org team to reply if those comments make sense.

dwradcliffe commented:
@simi has some good points. Smaller the changes the better.

sonalkr132 commented:
I have changed only `halt_callback_chains_on_return_false` config option because it was showing deprecation in test runs. That really irks me :persevere: I guess we can change it later.

I have removed commit which added rails version to migration as well because we don't *really* need that either. Migration deprecation show up only when migration are run..

homu commented:
:umbrella: The latest upstream changes (presumably #1385) made this pull request unmergeable. Please resolve the merge conflicts.

indirect commented:
@sonalkr132 can you re base this and update to 5.0.1 to fix the connection pool issue?

sonalkr132 commented:
Rails 5.0.1 is not released yet. [rails 5.0.1 Milestone](https://github.com/rails/rails/milestone/53)
Rails 5.0.0.1 was released and it was just a security update.

indirect commented:
@sonalkr132 ah, I see. 😞 I guess this is blocked until Rails 5.0.1 comes out, then.

homu commented:
:umbrella: The latest upstream changes (presumably #1398) made this pull request unmergeable. Please resolve the merge conflicts.

arthurnn commented:
I agree 100% with @dwradcliffe . We cannot ship one big update PR to upgrade Rails version. it is too risky.
We need to slowly introduce the changes in master, and make sure they run on both versions. Once we have master able to run on rails 4.2 and 5, and we test it out the move to 5

connorshea commented on test/functional/api/v1/profiles_controller_test.rb:
```diff
@@ -47,7 +47,7 @@ def response_body
         setup do
           @user.update(hide_email: true)
-          get :show, id: @user.handle, format: format
+          get :show, params: { id: @user.handle }, format: format
```
Changes like this should be done in a follow-up PR, in my opinion. This adds unnecessary review overhead to the Pull Request.

connorshea commented:
This would be _a lot_ smaller if we didn't bother fixing deprecation warnings in this PR. We should focus on doing that afterward.

connorshea commented on Gemfile:
```diff
@@ -16,7 +16,7 @@ gem 'daemons'
 gem 'delayed_job'
 gem 'delayed_job_active_record'
-gem 'doorkeeper'
+gem 'doorkeeper', '~> 4.0.0.rc4'
```
This can be backported (and can use the stable 4.0.0 version)

dwradcliffe replied:
This is actually gone in master now.

connorshea commented on Gemfile.lock:
```diff
@@ -176,50 +179,50 @@ GEM
       ast (~> 2.2)
-    paul_revere (2.0.0)
-      rails (~> 4.0)
+    paul_revere (2.1.0.rc1)
```
paul_revere seems to be [abandoned](https://github.com/thoughtbot/paul_revere/issues/35) from what I can tell, can we replace it with something else?

connorshea commented on app/controllers/api/v1/api_keys_controller.rb:
```diff
@@ -7,9 +7,9 @@ def show
       if current_user
         respond_to do |format|
-          format.any(:all) { render text: current_user.api_key }
+          format.any(:all) { render plain: current_user.api_key }
```
Replacing `render text` with `render plain` seems like an unnecessary deprecation warning fix? Should be a follow-up PR.

connorshea commented on config/initializers/filter_parameter_logging.rb:
```diff
@@ -0,0 +1,4 @@
+# Be sure to restart your server when you modify this file.
+
+# Configure sensitive parameters which will be filtered from the log file.
+Rails.application.config.filter_parameters += [:password]
```
Does this override an existing `filter_parameters` config?

connorshea commented on app/controllers/api/base_controller.rb:
```diff
@@ -1,7 +1,6 @@
 class Api::BaseController < ApplicationController
   before_action :doorkeeper_authorize!, if: :doorkeeper_token
   before_action :authenticate_with_oauth, if: :doorkeeper_token
-  skip_before_action :require_ssl
```
What's the explanation for this change?

connorshea commented on Gemfile.lock:
```diff
@@ -142,7 +144,7 @@ GEM
     gravtastic (3.2.6)
     hashie (3.4.4)
-    high_voltage (2.4.0)
+    high_voltage (3.0.0)
```
This seems to be something we can backport.

sonalkr132 commented:
Thanks for review @connorshea I will keep your comments in mind when I make changes.

> We cannot ship one big update PR to upgrade Rails version. it is too risky.

If it's okay I would like to open a new PR with :point_up:. It doesn't it has to wait for rails 5.0.1 for people to review it.

> We need to slowly introduce the changes in master, and make sure they run on both versions.

I haven't read (blog/implementation) anything related to this. Can someone please explain it with a little more detail? Would it be something like [this](https://github.com/rubygems/rubygems.org/pull/1262/files#diff-2812c7f41474174f2a338265b679244aR86), where we check for rails versions?

connorshea commented:
@sonalkr http://shayfrendt.com/posts/upgrading-github-to-rails-3-with-zero-downtime/

connorshea commented:
See also what I'm doing for GitLab: https://gitlab.com/gitlab-org/gitlab-ce/issues/14286

Lots of things can be backported to master to make this PR a lot smaller.

sonalkr132 commented:
@connorshea thanks :) That was helpful. I will send a new PR soon.

sonalkr132 commented:
See #1615
